File: #08.cpp
#include <iostream>
using namespace std;

class TreeNode {
public:
	int value;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int val) {
		value = val;
		left = nullptr;
		right = nullptr;
	}
};

// At each node must return whether tree under node is unival
// and the number of unival trees under
pair<bool, int> getUnivalTrees(TreeNode* t) {
	if (t->left == nullptr && t->right == nullptr) {
		return make_pair(true, 1);

	} else if (t->left == nullptr) {
		pair<bool, int> p = getUnivalTrees(t->right);

		// Check if current tree is unival
		if (p.first && t->value == t->right->value) return make_pair(true, p.second + 1);
		else return make_pair(false, p.second);

	} else if (t->right == nullptr) {
		pair<bool, int> p = getUnivalTrees(t->left);

		// Check if current tree is unival
		if (p.first && t->value == t->left->value) return make_pair(true, p.second + 1);
		else return make_pair(false, p.second);

	} else {
		pair<bool, int> left = getUnivalTrees(t->left);
		pair<bool, int> right = getUnivalTrees(t->right);

		if (left.first && right.first && t->value == t->right->value && t->value == t->left->value) {
			// Both branches and current tree are all unival
			return make_pair(true, left.second + right.second + 1);
		} else {
			// Sum number of unival subtrees from both branches
			return make_pair(false, left.second + right.second);
		}
	}
}

int main() {
	TreeNode* root = new TreeNode(0);
	root->left = new TreeNode(1);
	root->right = new TreeNode(0);
	root->right->left = new TreeNode(1);
	root->right->right = new TreeNode(0);
	root->right->left->left = new TreeNode(1);
	root->right->left->right = new TreeNode(1);

	cout << getUnivalTrees(root).second << endl;
}

File: #08.py

class TreeNode:
	def __init__(self, val):
		self.val = val
		self.left = None
		self.right = None


def get_unival_trees(t):
	if t.left == None and t.right == None:
		return (True, 1)

	elif t.left == None:
		(is_unival, num) = get_unival_trees(t.right)

		if is_unival and t.val == t.right.val:
			return (True, num + 1)
		else:
			return (False, num)

	elif t.right == None:
		(is_unival, num) = get_unival_trees(t.left)

		if is_unival and t.val == t.left.val:
			return (True, num + 1)
		else:
			return (False, num)

	else:
		(is_unival_left, num_left) = get_unival_trees(t.left)
		(is_unival_right, num_right) = get_unival_trees(t.right)

		if is_unival_right and is_unival_left and t.val == t.left.val and t.val == t.right.val:
			return (True, num_right + num_left + 1)
		else:
			return (False, num_left + num_right)


root = TreeNode(0)
root.left = TreeNode(1)
root.right = TreeNode(0)
root.right.left = TreeNode(1)
root.right.left.left = TreeNode(1)
root.right.left.right = TreeNode(1)
root.right.right = TreeNode(0)

print(get_unival_trees(root)[1])
